src/libs/kata-sys-util/src/fs.rs

```diff
@@ -0,0 +1,234 @@
+    let path = path.as_ref();
+    let meta = fs::symlink_metadata(path)?;
+
+    Ok(meta.file_type().is_symlink())
+}
+
+/// Reflink copy src to dst, and falls back to regular copy if reflink copy fails.
+///
+/// # Safety
+/// The `reflink_copy()` doesn't preserve permission/security context for the copied file,
+/// so caller needs to take care of it.
+pub fn reflink_copy<S: AsRef<Path>, D: AsRef<Path>>(src: S, dst: D) -> Result<()> {
+    let src_path = src.as_ref();
+    let dst_path = dst.as_ref();
+    let src = src_path.to_string_lossy();
+    let dst = dst_path.to_string_lossy();
+
+    if !src_path.is_file() {
+        return Err(eother!("reflink_copy src {} is not a regular file", src));
+    }
+
+    // Make sure dst's parent exist. If dst is a regular file, then unlink it for later copy.
+    if dst_path.exists() {
+        if !dst_path.is_file() {
+            return Err(eother!("reflink_copy dst {} is not a regular file", dst));
+        } else {
+            fs::remove_file(dst_path)?;
+        }
+    } else if let Some(dst_parent) = dst_path.parent() {
+        if !dst_parent.exists() {
+            if let Err(e) = fs::create_dir_all(dst_parent) {
+                return Err(eother!(
+                    "reflink_copy: create_dir_all {} failed: {:?}",
+                    dst_parent.to_str().unwrap(),
+                    e
+                ));
+            }
+        } else if !dst_parent.is_dir() {
+            return Err(eother!("reflink_copy parent of {} is not a directory", dst));
+        }
+    }
+
+    // Reflink copy, and fallback to regular copy if reflink fails.
+    let src_file = fs::File::open(src_path)?;
+    let dst_file = fs::File::create(dst_path)?;
+    if let Err(e) = do_reflink_copy(src_file, dst_file) {
+        match e.raw_os_error() {
+            // Cross dev copy or filesystem doesn't support reflink, do regular copy
+            Some(os_err)
+                if os_err == nix::Error::EXDEV as i32
+                    || os_err == nix::Error::EOPNOTSUPP as i32 =>
+            {
+                warn!(
+                    sl!(),
+                    "reflink_copy: reflink is not supported ({:?}), do regular copy instead", e,
+                );
+                if let Err(e) = do_regular_copy(src.as_ref(), dst.as_ref()) {
+                    return Err(eother!(
+                        "reflink_copy: regular copy {} to {} failed: {:?}",
+                        src,
+                        dst,
+                        e
+                    ));
+                }
+            }
+            // Reflink copy failed
+            _ => {
+                return Err(eother!(
+                    "reflink_copy: copy {} to {} failed: {:?}",
+                    src,
+                    dst,
+                    e,
+                ))
+            }
+        }
+    }
+
+    Ok(())
+}
+
+// Copy file using cp command, which handles sparse file copy.
+fn do_regular_copy(src: &str, dst: &str) -> Result<()> {
```

Comment:
Should these parameters be checked? See my questions for `reflink_copy()`.

Reply:
It's a private function, and it depends on `reflink_copy()` to do the protection.